### PR TITLE
metadata: init at 0.1.9

### PR DESCRIPTION
**Pull request by myclevorname**
## Description of changes

Info about `metadata` can be found at #322243.

Hopefully, `metadata` should build, but requires libavutil, whose request is #322768.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/me/metadata/package.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  nativeBuildInputs = [ pkg-config asciidoc ];
+
+  postBuildHook = "a2x --doctype manpage --format manpage man/metadata.1.adoc";
+  postInstallHook = ''
+    mkdir -p $out/share/man
+    cp man/metadata.1 $out/share/man/
+  '';
+
+  # Run-time AND build-time
+  BuildInputs = [ ffmpeg libavutil ];
```
Comment (pinpox) [inline pkgs/by-name/me/metadata/package.nix:25]:
```suggestion
  buildInputs = [ ffmpeg libavutil ];
```
`buidInputs` is lowercase. Also unnessary explanation, maintainers should know what this does.

File: pkgs/by-name/me/metadata/package.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  src = fetchFromGitHub {
+    owner = "zmwangx";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-OFWdCV9Msy/mNaSubqoJi4tBiFqL7RuWWQluSnKe4fU=";
+  };
+
+  cargoHash = "sha256-F5jXS/W600nbQtu1FD4+DawrFsO+5lJjvAvTiFKT840=";
+
+  # Build-time
```
Comment (pinpox) [inline pkgs/by-name/me/metadata/package.nix:15]:
```suggestion
```

File: pkgs/by-name/me/metadata/package.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
+{ stdenv, lib, fetchFromGitHub, pkg-config, ffmpeg, rustPlatform, asciidoc, libavutil }:
```
Comment (pinpox) [inline pkgs/by-name/me/metadata/package.nix:1]: `libavutil` does not exist here.

**Review by pinpox — CHANGES_REQUESTED** [review]

**Comment by pinpox** [comment]
I tried playing around with this, this package should work:

```nix
{ clang
, libclang
, stdenv
, lib
, fetchFromGitHub
, pkg-config
, ffmpeg
, rustPlatform
, asciidoc
, glib
, llvmPackages
}:
rustPlatform.buildRustPackage rec {
  pname = "metadata";
  version = "0.1.9";

  src = fetchFromGitHub {
    owner = "zmwangx";
    repo = pname;
    rev = "v${version}";
    hash = "sha256-OFWdCV9Msy/mNaSubqoJi4tBiFqL7RuWWQluSnKe4fU=";
  };

  cargoHash = "sha256-F5jXS/W600nbQtu1FD4+DawrFsO+5lJjvAvTiFKT840=";

  nativeBuildInputs = [
    pkg-config
    asciidoc
    clang
    libclang.lib
    llvmPackages.libcxxClang
    clang
  ];

  postBuildHook = "a2x --doctype manpage --format manpage man/metadata.1.adoc";
  postInstallHook = ''
    mkdir -p $out/share/man
    cp man/metadata.1 $out/share/man/
  '';

  buildInputs = [
    ffmpeg
    clang
    glib
    llvmPackages.libclang
    stdenv.cc.libc
  ];

  env.LIBCLANG_PATH = "${libclang.lib}/lib";
  env.FFMPEG_DIR = ffmpeg.dev;

  meta = {
    description = "Media file metadata for human consumption";
    longDescription = ''
      metadata is a media metadata parser and formatter designed for human consumption.
      Powered by FFmpeg.
    '';
    maintainers = with lib.maintainers; [ clevor ];
    license = lib.licenses.mit;
    homepage = "https://github.com/zmwangx/metadata";
    platforms = lib.platforms.linux;
    mainProgram = "metadata";
  };
}
```

Compilation will still fail though, because of this issue which has to be fixed upstream: https://github.com/zmwangx/metadata/issues/9

**Comment by pinpox** [comment]
Result of `nixpkgs-review pr 322773` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>metadata</li>
  </ul>
</details>

**Review by pinpox — APPROVED** [review]

File: pkgs/by-name/me/metadata/package.nix [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    clang
+    libclang.lib
+    llvmPackages.libcxxClang
+    clang
+  ];
+
+  postBuildHook = "a2x --doctype manpage --format manpage man/metadata.1.adoc";
+  postInstallHook = ''
+    mkdir -p $out/share/man
+    cp man/metadata.1 $out/share/man/
```
Comment (Aleksanaa) [inline pkgs/by-name/me/metadata/package.nix:38]: Please use `installManpage`: `installManpage man/metadata.1`. No need to mkdir.

File: pkgs/by-name/me/metadata/package.nix [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  '';
+
+  buildInputs = [
+    ffmpeg_7
+    clang
+    glib
+    llvmPackages.libclang
+    stdenv.cc.libc
```
Comment (Aleksanaa) [inline pkgs/by-name/me/metadata/package.nix:46]:
```suggestion
```

File: pkgs/by-name/me/metadata/package.nix [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+  cargoHash = "sha256-F5jXS/W600nbQtu1FD4+DawrFsO+5lJjvAvTiFKT840=";
+
+  nativeBuildInputs = [
+    pkg-config
+    asciidoc
+    clang
+    libclang.lib
+    llvmPackages.libcxxClang
+    clang
```
Comment (Aleksanaa) [inline pkgs/by-name/me/metadata/package.nix:32]: Please try adding `rustPlatform.bindgenHook` and removing all "clang" stuff.

File: pkgs/by-name/me/metadata/package.nix [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    glib
+    llvmPackages.libclang
+    stdenv.cc.libc
+  ];
+
+  env.LIBCLANG_PATH = "${libclang.lib}/lib";
+  env.FFMPEG_DIR = ffmpeg_7.dev;
+
+  meta = {
+    description = "Media file metadata for human consumption";
```
Comment (Aleksanaa) [inline pkgs/by-name/me/metadata/package.nix:53]:
```suggestion
    description = "Media metadata parser and formatter designed for human consumption, powered by FFmpeg";
```

No need to add `longDescription`.

File: pkgs/by-name/me/metadata/package.nix [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    longDescription = ''
+      metadata is a media metadata parser and formatter designed for human consumption.
+      Powered by FFmpeg.
+    '';
+    maintainers = with lib.maintainers; [ clevor ];
+    license = lib.licenses.mit;
+    homepage = "https://github.com/zmwangx/metadata";
+    platforms = lib.platforms.linux;
```
Comment (Aleksanaa) [inline pkgs/by-name/me/metadata/package.nix:61]:
```suggestion
```

`rustPlatform` comes with a platform by default. Seems this is also available on macOS.

File: pkgs/by-name/me/metadata/package.nix [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+, glib
+, llvmPackages
+}:
+rustPlatform.buildRustPackage rec {
+  pname = "metadata";
+  version = "0.1.9";
+
+  src = fetchFromGitHub {
+    owner = "zmwangx";
+    repo = pname;
```
Comment (Aleksanaa) [inline pkgs/by-name/me/metadata/package.nix:19]:
```suggestion
    repo = "metadata";
```

File: pkgs/by-name/me/metadata/package.nix (hunk 4 above)
Comment (Aleksanaa) [reply pkgs/by-name/me/metadata/package.nix:38]: It's in `installShellFiles`. Add this to `nativeBuildInputs` first.

File: pkgs/by-name/me/metadata/package.nix [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+  cargoHash = "sha256-F5jXS/W600nbQtu1FD4+DawrFsO+5lJjvAvTiFKT840=";
+
+  nativeBuildInputs = [
+    pkg-config
+    asciidoc
+  ];
+
+  postBuildHook = "a2x --doctype manpage --format manpage man/metadata.1.adoc";
+  postInstallHook = ''
```
Comment (Aleksanaa) [inline pkgs/by-name/me/metadata/package.nix:31]:
```suggestion
  postInstall = ''
```

File: pkgs/by-name/me/metadata/package.nix [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  };
+
+  cargoHash = "sha256-F5jXS/W600nbQtu1FD4+DawrFsO+5lJjvAvTiFKT840=";
+
+  nativeBuildInputs = [
+    pkg-config
+    asciidoc
+  ];
+
+  postBuildHook = "a2x --doctype manpage --format manpage man/metadata.1.adoc";
```
Comment (Aleksanaa) [inline pkgs/by-name/me/metadata/package.nix:30]:
```suggestion
  postBuild = ''
    a2x --doctype manpage --format manpage man/metadata.1.adoc
  '';

```

File: pkgs/by-name/me/metadata/package.nix (hunk 6 above)
Comment (Aleksanaa) [reply pkgs/by-name/me/metadata/package.nix]: It should be in `nativeBuildInputs`

Comment (Aleksanaa) [reply pkgs/by-name/me/metadata/package.nix]: Try removing that LIBCLANG after you apply this change. It should work

**Review by Aleksanaa — DISMISSED** [review]
Please try moving `rustPlatform.bindgenHook` to `nativeBuildInputs` and removing `LIBCLANG_PATH`.

**Comment by pinpox** [comment]
Result of `nixpkgs-review pr 322773` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>metadata</li>
  </ul>
</details>

**Review by pinpox — APPROVED** [review]

**Comment by Aleksanaa** [comment]
Have to pin to a higher darwin toolchain stdenv? Not sure

**Comment by Aleksanaa** [comment]
```
libavdevice/avfoundation.m:765:41: error: must use 'enum' tag to refer to type 'AVMediaType'
static NSArray* getDevicesWithMediaType(AVMediaType mediaType) {
                                        ^
                                        enum 
libavdevice/avfoundation.m:816:50: error: incompatible integer to pointer conversion sending 'enum AVMediaType' to parameter of type 'NSString *' [-Wint-conversion]
    return [AVCaptureDevice devicesWithMediaType:mediaType];
                                                 ^~~~~~~~~
/nix/store/aq7sq7iwahdkzxn00ksxhpx81pd3aaw3-apple-framework-AVFoundation/Library/Frameworks/AVFoundation.framework/Headers/AVCaptureDevice.h:110:47: note: passing argument to parameter 'mediaType' here
+ (NSArray *)devicesWithMediaType:(NSString *)mediaType;
                                              ^
libavdevice/avfoundation.m:829:54: error: incompatible pointer to integer conversion passing 'NSString *const' to parameter of type 'enum AVMediaType' [-Wint-conversion]
    NSArray *devices       = getDevicesWithMediaType(AVMediaTypeVideo);
                                                     ^~~~~~~~~~~~~~~~
libavdevice/avfoundation.m:765:53: note: passing argument to parameter 'mediaType' here
static NSArray* getDevicesWithMediaType(AVMediaType mediaType) {
                                                    ^
libavdevice/avfoundation.m:830:54: error: incompatible pointer to integer conversion passing 'NSString *const' to parameter of type 'enum AVMediaType' [-Wint-conversion]
    NSArray *devices_muxed = getDevicesWithMediaType(AVMediaTypeMuxed);
                                                     ^~~~~~~~~~~~~~~~
libavdevice/avfoundation.m:765:53: note: passing argument to parameter 'mediaType' here
static NSArray* getDevicesWithMediaType(AVMediaType mediaType) {
                                                    ^
libavdevice/avfoundation.m:865:43: error: incompatible pointer to integer conversion passing 'NSString *const' to parameter of type 'enum AVMediaType' [-Wint-conversion]
        devices = getDevicesWithMediaType(AVMediaTypeAudio);
                                          ^~~~~~~~~~~~~~~~
libavdevice/avfoundation.m:765:53: note: passing argument to parameter 'mediaType' here
static NSArray* getDevicesWithMediaType(AVMediaType mediaType) {
                                                    ^
```

**Comment by wegank** [comment]
Not the fault of this PR: https://hydra.nixos.org/job/nixpkgs/trunk/ffmpeg_7.x86_64-darwin

**Comment by Aleksanaa** [comment]
Sorry, didn't look up the logs carefully

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**